### PR TITLE
ISPN-13977 Create HR 4.0 protocol

### PR DESCRIPTION
**Pull request by wburns**
https://issues.redhat.com/browse/ISPN-13977

* Bump HR protocol to 4.0
* FORCE_RETURN_VALUE is no longer supported with new API (dictated solely by API usage)
* Change return values on various API methods to return CacheEntry<K, V> instead of just V
* Server protocol 4.0 support to return entry instead of just value
* Change old client to work with both entry and value return based on protocol version
* Add in dummy server code to return entries (still needs to have proper implementations added)

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec40.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+   }
+
+   @Override
+   protected HeaderParams writeHeader(ByteBuf buf, HeaderParams params, byte version) {
+      HeaderParams headerParams = super.writeHeader(buf, params, version);
+      writeOtherParams(buf, params.otherParams());
+      return headerParams;
+   }
+
+   private void writeOtherParams(ByteBuf buf, Map<String, String> parameters) {
```
Comment (pruivo) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec40.java:41]: I commented on Fabio's PR... I think we decided to make the value a `byte[]`

File: client/hotrod/src/main/java/org/infinispan/hotrod/impl/operations/PutIfAbsentOperation.java [hunk 2]
```diff
@@ -1,52 +1,45 @@
...
-         if (log.isTraceEnabled()) {
-            log.tracef("Returning from putIfAbsent: %s", prevValue);
-         }
-         complete(prevValue);
-      } else {
-         statsDataStore();
-         complete(null);
-      }
+   protected int flags() {
+      return super.flags() | PrivateHotRodFlag.FORCE_RETURN_VALUE.getFlagInt();
```
Comment (pruivo) [inline client/hotrod/src/main/java/org/infinispan/hotrod/impl/operations/PutIfAbsentOperation.java:43]: shouldn't the private flags be shiffed? like half of the int for public flags and the other half for private flags?

File: core/src/main/java/org/infinispan/AdvancedCache.java [hunk 3]
```diff
@@ -330,6 +337,12 @@ default CompletableFuture<V> replaceAsync(K key, V value, Metadata metadata) {
       return replaceAsync(key, value, metadata.lifespan(), TimeUnit.MILLISECONDS, metadata.maxIdle(), TimeUnit.MILLISECONDS);
    }
 
+   default CompletionStage<CacheEntry<K, V>> replaceAsyncReturnEntry(K key, V value, Metadata metadata) {
+      // TODO: implement later
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/AdvancedCache.java:341]: I'll assume there is a JIRA for this :)

**Review by pruivo — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/infinispan/AdvancedCache.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/AdvancedCache.java:341]: @jabolina is working on the embedded versions of this, yes. The new API doesn't work until these are implemented. However, the old API works just fine. So I was just leaving them separate to not make this PR that big.

File: client/hotrod/src/main/java/org/infinispan/hotrod/impl/operations/PutIfAbsentOperation.java (hunk 2 above)
Comment (wburns) [reply client/hotrod/src/main/java/org/infinispan/hotrod/impl/operations/PutIfAbsentOperation.java:43]: I didn't want to change the numeric value of the flag to have easier backwards compatibility.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec40.java (hunk 1 above)
Comment (wburns) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec40.java:41]: Sure, I would say any changes in there @fax4ever can let me know and I can squash any of his review changes

File: api/src/main/java/org/infinispan/api/async/AsyncCache.java [hunk 4]
```diff
@@ -157,7 +161,10 @@ default CompletionStage<Void> set(K key, V value) {
     * @param options
     * @return
     */
-   CompletionStage<Void> set(K key, V value, CacheWriteOptions options);
+   default CompletionStage<Void> set(K key, V value, CacheWriteOptions options) {
+      return put(key, value, options)
+            .thenApply(__ -> null);
```
[the remark this answers is not in the document]
Comment (wburns) [reply api/src/main/java/org/infinispan/api/async/AsyncCache.java:166]: Sure, although the final goal is to eventually not need these.

**Comment by wburns** [comment]
Added a commit that should hopefully fix the failures

**Comment by fax4ever** [comment]
@wburns I updated my contribution for https://github.com/infinispan/infinispan/pull/10202.
I changed the `Map<String, String>` => `Map<String, byte[]>`

File: core/src/main/java/org/infinispan/AdvancedCache.java (hunk 3 above)
Comment (jabolina) [reply core/src/main/java/org/infinispan/AdvancedCache.java:341]: Created #10207 for these TODOs.

**Comment by wburns** [comment]
Added https://github.com/infinispan/infinispan/pull/10196/files#diff-129dcbfa1d2f2d809cb69dd2352a598f5ea9e68bee9c4a317f0dd8efe6ff6759 which tests all the protocol versions with return values.

**Comment by tristantarrant** [comment]
Two related failures:
getPut – org.infinispan.hotrod.HotRodAsyncCacheTest1m 30s
testCaches – org.infinispan.hotrod.HotRodAsyncCachesTest

**Comment by wburns** [comment]
> Two related failures: getPut – org.infinispan.hotrod.HotRodAsyncCacheTest1m 30s testCaches – org.infinispan.hotrod.HotRodAsyncCachesTest

Confirmed, checking.

**Comment by wburns** [comment]
Pushed with newest client changes, tests should pass now 🤞

**Comment by tristantarrant** [comment]
Merged, thanks

**Pull request merged**